crates/primitives-traits/src/node.rs

```diff
@@ -78,33 +79,73 @@ impl NodePrimitives for () {
 
 /// Helper trait that sets trait bounds on [`NodePrimitives`].
 pub trait FullNodePrimitives:
-    Send + Sync + Unpin + Clone + Default + fmt::Debug + PartialEq + Eq + 'static
+    NodePrimitives<
+    Block: Block<Header = Self::BlockHeader, Body = Self::BlockBody> + 'static,
+    BlockHeader: BlockHeader + 'static,
+    BlockBody: BlockBody<Transaction = Self::SignedTx> + 'static,
+    SignedTx: FullSignedTx + 'static,
+    TxType: FullTxType + 'static,
+    Receipt: FullReceipt + 'static,
+>
 {
-    /// Block primitive.
-    type Block: FullBlock<Header = Self::BlockHeader, Body = Self::BlockBody>;
-    /// Block header primitive.
-    type BlockHeader: FullBlockHeader + 'static;
-    /// Block body primitive.
-    type BlockBody: FullBlockBody<Transaction = Self::SignedTx> + 'static;
-    /// Signed version of the transaction type.
-    type SignedTx: FullSignedTx;
-    /// Transaction envelope type ID.
-    type TxType: FullTxType;
-    /// A receipt.
-    type Receipt: FullReceipt;
 }
 
-impl<T> NodePrimitives for T
-where
-    T: FullNodePrimitives,
+impl<T> FullNodePrimitives for T where
+    T: NodePrimitives<
+        Block: FullBlock<Header = Self::BlockHeader, Body = Self::BlockBody> + 'static,
+        BlockHeader: FullBlockHeader + 'static,
+        BlockBody: FullBlockBody<Transaction = Self::SignedTx> + 'static,
+        SignedTx: FullSignedTx + 'static,
+        TxType: FullTxType + 'static,
+        Receipt: FullReceipt + 'static,
+    >
 {
-    type Block = T::Block;
-    type BlockHeader = T::BlockHeader;
-    type BlockBody = T::BlockBody;
-    type SignedTx = T::SignedTx;
-    type TxType = T::TxType;
-    type Receipt = T::Receipt;
 }
 
 /// Helper adapter type for accessing [`NodePrimitives`] receipt type.
 pub type ReceiptTy<N> = <N as NodePrimitives>::Receipt;
+
+/// Helper trait that links [`NodePrimitives::Block`] to [`NodePrimitives::BlockHeader`],
+/// [`NodePrimitives::BlockBody`] and [`NodePrimitives::SignedTx`].
+pub trait BlockPrimitives:
```

Comment:
unsure I understand why we need a separate trait for this? those bounds already exist for `FullNodePrimitives`, and I believe we could also move them to `NodePrimitives`?

Reply:
for the same reason we don't have the compact feature `reth-codec` always enabled, cc @joshieDo . it makes testing more straight forward and it makes it more apparent to us how data flows in reth.

if you want to test for example consensus engine, you should just need an impl like
```rust
impl NodePrimitives for TestPrimitives {
    type Block = Block;
    type BlockHeader = Header;
    type BlockBody = BlockBody;
    type SignedTx = TransactionSigned;
    type TxType = ();
    type Receipt = ();
}
```

it's always misleading a weird for someone new to the code, to try and figure out how the code works by looking at the test, and then realise half way through that types are in scope, that are not used.

Reply:
nice side effect of having `NodePrimitives` as super trait of `BlockPrimitives`, is that in case it's needed, a trait impl using `BlockPrimitives`, can still return a receipt type, just not use the receipt functionality in its scope

Reply:
`ProviderFactory` won't work unless primitives are `FullNodePrimitives` so I don't think you will be able to test anything with such primitives implementation

Reply:
I think it might make sense to move those bounds to `NodePrimitives` but separate trait doesn't seem very useful

Reply:
provider is expected to use `FullNodePrimitives` in some places. node builder is expected to use `FullNodePrimitives` everywhere